src/lib/ams.c

```diff
@@ -0,0 +1,736 @@
+
+		if (ixc_route == cpu_id) {
+			/* we are on target core already */
+			/* release lock here, callback are NOT supposed to change routing_table */
+			ams_release(shared_c);
+
+			ams_target.consumer_callback(ams_message_payload, ams_target.ctx);
+			err = 0;
+		} else {
+			/* we have to go through idc */
+			if (incoming) {
+				/* if bit is set we are forwarding it again */
+				if (shared_c->slot_done[incoming_slot] & BIT(ams_target.consumer_core_id)) {
+					/* slot was already processed for that core, skip it */
+					continue;
+				}
+			}
+			/* if bit clear */
+			if ((forwarded & BIT(ams_target.consumer_core_id)) == 0) {
+				/* we have consumer for previously untouched core */
+
+				/* bump uses count, mark current as processed already */
+				if (slot != AMS_INVALID_SLOT) {
+					shared_c->slot_uses[slot]++;
+					shared_c->slot_done[slot] |= BIT(cpu_id);
+				}
+
+				/* release lock here, so other core can acquire it again */
+				ams_release(shared_c);
+
+				if (slot != AMS_INVALID_SLOT) {
+					forwarded |= BIT(cpu_id);
+					err = ams_send_over_ixc(ams, slot, &ams_target);
+					if (err != 0) {
+						/* idc not sent, update slot refs locally */
+						shared_c = ams_acquire(ams->ams_context->shared);
+						shared_c->slot_uses[slot]--;
+						shared_c->slot_done[slot] |= BIT(ams_target.consumer_core_id);
+						ams_release(shared_c);
+					}
+				}
+			} else {
+				/* should not happen but skip reacquiring just in case */
+				continue;
+			}
+		}
+
+		/* acquire shared context lock again */
+		shared_c = ams_acquire(ams->ams_context->shared);
+	}
+
+	if (incoming)
+		shared_c->slot_uses[incoming_slot]--;
+
+	ams_release(shared_c);
+
+	if (!found_any)
+		tr_err(&ams_tr, "No entries found!");
+
+	return err;
+}
+
+int ams_send(const struct ams_message_payload *const ams_message_payload)
+{
+	struct async_message_service *ams = sof_get()->ams;
+
+	return ams_message_send_internal(ams, ams_message_payload, AMS_ANY, AMS_ANY,
+					 AMS_INVALID_SLOT);
+}
+
+int ams_message_send_mi(struct async_message_service *ams,
+			const struct ams_message_payload *const ams_message_payload,
+			uint16_t target_module, uint16_t target_instance)
+{
+	return ams_message_send_internal(ams, ams_message_payload, target_module,
+					 target_instance, AMS_INVALID_SLOT);
+}
+
+int ams_send_mi(const struct ams_message_payload *const ams_message_payload,
+		uint16_t module_id, uint16_t instance_id)
+{
+	struct async_message_service *ams = sof_get()->ams;
+
+	return ams_message_send_mi(ams, ams_message_payload, module_id, instance_id);
+}
+
+static int send_message_over_ixc(struct async_message_service *ams, uint32_t slot,
+				 struct ams_routing_entry *target)
+{
+	if (target == NULL)
+		return -EINVAL;
+
+	int ixc_route = ams_get_ixc_route_to_target(cpu_get_id(),
+							 target->consumer_core_id);
+
+	struct idc_msg ams_request = {
+		.header = IDC_MSG_AMS | slot,
+		.extension = IDC_MSG_AMS_EXT,
+		.core = ixc_route,
```

Comment:
if I understand it correctly, say, if core 0 wants to send a message to core 1, this function will be called with `ixc_route == 1` right? Then `idc_cmd()` will run on core 1, which will call `idc_process_async_msg()` and then `process_incoming_message()` and then `ams_task_add_slot_to_process()` so, the latter can run on secondary cores too?